### PR TITLE
Add syntax ALTER USER MODIFY SETTINGS

### PR DESCRIPTION
**Pull request by vitlibar**
### Changelog category:
- Improvement

### Changelog entry:
Add syntax `ALTER USER {ADD|MODIFY|DROP SETTING}`, `ALTER USER {ADD|DROP PROFILE}`, the same for `ALTER ROLE` and `ALTER PROFILE`.

---
READY FOR REVIEW

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit dc983fc2e004b3b48769cf6bbfee0662b574a758 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57421/dc983fc2e004b3b48769cf6bbfee0662b574a758/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/57421/dc983fc2e004b3b48769cf6bbfee0662b574a758/ci_running.html">✅ success</a></td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/7060626952">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Comment by nikitamikhaylov** [comment]
@pufit The PR description says it is ready for review ;)

**Comment by vitlibar** [comment]
@pufit It seems you wanted to review this PR ;)

**Comment by pufit** [comment]
> @pufit It seems you wanted to review this PR ;)

Sorry, I've started looking into it and totally forgot to finish. Will post my thoughts by the end of the day

File: src/Access/SettingsProfileElement.cpp [hunk 1]
```diff
@@ -236,6 +253,81 @@ std::vector<UUID> SettingsProfileElements::toProfileIDs() const
...
+            }
+        }
+        insert(end(), new_elements.begin(), new_elements.end());
+    }
+
+    /// Partitioning: first profiles, then settings.
+    /// We use std::stable_partition() here because we want to preserve the relative order of profiles and the relative order of settings.
+    /// (We need that order to be preserved to remove duplicates correctly - see below.)
+    auto profiles_begin = begin();
+    auto profiles_end = std::stable_partition(begin(), end(), [](const SettingsProfileElement & element) { return static_cast<bool>(element.parent_profile); });
```
Comment (pufit) [inline src/Access/SettingsProfileElement.cpp:279]: I understand the idea behind this, but it still feels wrong that we have two different entities (settings and profiles) in one vector and do this tricky reordering each time. Maybe we should separate it somehow in the first place?

File: src/Access/SettingsConstraints.cpp [hunk 2]
```diff
@@ -133,6 +133,13 @@ void SettingsConstraints::merge(const SettingsConstraints & other)
 }
 
 
+void SettingsConstraints::check(const Settings & current_settings, const AlterSettingsProfileElements & profile_elements, SettingSource source) const
+{
+    check(current_settings, profile_elements.add_settings, source);
+    check(current_settings, profile_elements.modify_settings, source);
+    /// We don't check `drop_settings` here.
```
Comment (pufit) [inline src/Access/SettingsConstraints.cpp:140]: Why?

File: src/Parsers/Access/ASTSettingsProfileElement.cpp [hunk 3]
```diff
@@ -82,6 +129,35 @@ bool ASTSettingsProfileElements::empty() const
     return true;
 }
 
+size_t ASTSettingsProfileElements::getNumberOfSettings() const
+{
+    size_t count = 0;
```
Comment (pufit) [inline src/Parsers/Access/ASTSettingsProfileElement.cpp:134]: std::count_if?

File: src/Parsers/Access/ASTSettingsProfileElement.cpp [hunk 4]
```diff
@@ -82,6 +129,35 @@ bool ASTSettingsProfileElements::empty() const
...
+    for (const auto & element : elements)
+        if (!element->setting_name.empty())
+            ++count;
+    return count;
+}
+
+size_t ASTSettingsProfileElements::getNumberOfProfiles() const
+{
+    size_t count = 0;
```
Comment (pufit) [inline src/Parsers/Access/ASTSettingsProfileElement.cpp:143]: std::count_if?

**Comment by pufit** [comment]
https://github.com/ClickHouse/ClickHouse/pull/72050

**Pull request closed without merging**